Software/app/freefall_lorawan/lora_app.c

```diff
@@ -0,0 +1,301 @@
+  }
+}
+
+void HAL_GPIO_EXTI_Callback(uint16_t GPIO_Pin)
+{
+  if (GPIO_Pin == ACC_INT_PIN)
+  {
+    UTIL_SEQ_SetTask((1 << CFG_SEQ_Task_LoRaSendOnFreefall), CFG_SEQ_Prio_0);
+  }
+}
+
+static void OnRxData(LmHandlerAppData_t *appData, LmHandlerRxParams_t *params)
+{
+  if ((appData != NULL) && (params != NULL))
+  {
+    static const char *slotStrings[] = {"1", "2", "C", "C Multicast", "B Ping-Slot", "B Multicast Ping-Slot"};
+
+    APP_LOG(TS_OFF, VLEVEL_M, "\r\n ###### ========== MCPS-Indication ==========\r\n");
+    APP_LOG(TS_OFF, VLEVEL_M, "\r\n ###### D/L FRAME:%04d | SLOT:%s | PORT:%d | DR:%d | RSSI:%d | SNR:%d\r\n",
+            params->DownlinkCounter, slotStrings[params->RxSlot], appData->Port, params->Datarate, params->Rssi, params->Snr);
+    switch (appData->Port)
+    {
+    case LRAWAN_APP_SWITCH_CLASS_PORT:
+      /*this port switches the class*/
+      if (appData->BufferSize == 1)
+      {
+        switch (appData->Buffer[0])
+        {
+        case LRAWAN_APP_SWITCH_CLASS_A:
+        {
+          LmHandlerRequestClass(CLASS_A);
+          break;
+        }
+        case LRAWAN_APP_SWITCH_CLASS_B:
+        {
+#if defined(LORAMAC_CLASSB_ENABLED) && (LORAMAC_CLASSB_ENABLED == 1)
+          LmHandlerRequestClass(CLASS_B);
+#else
+          APP_LOG(TS_OFF, VLEVEL_M, "\r\n Configure LORAMAC_CLASSB_ENABLED to be able to switch to this class \r\n");
+#endif
+          break;
+        }
+        case LRAWAN_APP_SWITCH_CLASS_C:
+        {
+          LmHandlerRequestClass(CLASS_C);
+          break;
+        }
+        default:
+          break;
+        }
+      }
+      break;
+    case LORAWAN_APP_PORT:
+      APP_LOG(TS_OFF, VLEVEL_M, "\r\n Recieved %d bytes on LORAWAN_APP_PORT: %d \r\n", appData->BufferSize, LORAWAN_APP_PORT);
+      break;
+    default:
+      APP_LOG(TS_OFF, VLEVEL_M, "\r\n Recieved %d bytes on undefined port: %d \r\n", appData->BufferSize, LORAWAN_APP_PORT);
+      break;
+    }
+  }
+}
+
+static void SendTxData(void)
+{
+  static uint8_t freefall_log_amount;
+  UTIL_TIMER_Time_t nextTxIn = 0;
+
+  UTIL_TIMER_Create(&TxLedTimer, 0xFFFFFFFFU, UTIL_TIMER_ONESHOT, OnTimerLedEvent, NULL);
+  UTIL_TIMER_SetPeriod(&TxLedTimer, 200);
+
+  // User can add any indication here (LED manipulation or Buzzer)
+
+  UTIL_TIMER_Start(&TxLedTimer);
+
+  AppData.Port = LORAWAN_APP_PORT;
+  AppData.BufferSize = 1;
+  AppData.Buffer[0] = freefall_log_amount;
+
+  if (LORAMAC_HANDLER_SUCCESS == LmHandlerSend(&AppData, LORAWAN_DEFAULT_CONFIRMED_MSG_STATE, &nextTxIn, false))
+  {
+    APP_LOG(TS_ON, VLEVEL_L, "SEND REQUEST\r\n");
+    freefall_log_amount = 0;
+  }
+  else if (nextTxIn > 0)
+  {
+    APP_LOG(TS_ON, VLEVEL_L, "Next Tx in  : ~%d second(s)\r\n", (nextTxIn / 1000));
+    freefall_log_amount++;
+
+    // Set timer for next available transmission
+    UTIL_TIMER_Create(&TxTimer, 0xFFFFFFFFU, UTIL_TIMER_ONESHOT, OnTxTimerRetransmission, NULL);
+    UTIL_TIMER_SetPeriod(&TxTimer, nextTxIn);
+    UTIL_TIMER_Start(&TxTimer);

```

Comment:
@marnixcrouzen I think the logic here is incorrect, you are scheduling an event every time there is an interrupt and nextTxIn > 0

Reply:
@elsalahy These events are scheduled in case the transmission time was too low. The timer sets a new event for the next time a transmission is available. Otherwise the transmission would simply fail and it would wait for another free fall event (missing the failed attempt).